### ISSUE TITLE
训练集和测试集一样，修改dataprovider，用classification_error_evaluator 计算的训练集和测试集的准确率差距太大，于预期不符

### ISSUE THREAD

**Issue by lazzylee** [unisound-ail/Paddle#4]
我们希望搭建一个最简单的 NN demo，属于语音识别声学建模中的一个简单小模块。希望利用 NN 建模，达到提高特征分类的正确率。
使用 1000 句语音进行训练，同样的 1000 句语音进行测试。
484 维特征为一个 sample，每个 sample 都有一个预先准备好的 label，总计大约 260k sample pairs。
使用简单的 两层的全连接神经网络（fc_layer） 训练，每一层节点数相同，节点数尝试过 512，1024。输入特征维数为 484 维，输出类别数为3513。
setting 都使用文档中默认的参数。batch_size = 256, learning_rate=2e-3, learning_method=AdamOptimizer(), regularization=L2Regularization(8e-4), gradient_clipping_threshold=25
由于是特定的特征，我们修改了PyDataProvider2，使用[dataprovider_ark.py](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/dataprovider_ark.py)，当使用默认的@provider(init_hook=initializer)时，训练集和测试集的 classification_error_evaluator 都很低，大约 4%。（由于集内训练测试的缘故）
当使用 @provider(init_hook=initializer, cache=CacheType.CACHE_PASS_IN_MEM) 时，最终结果正常，与使用默认参数情况一样。
但当使用 @provider(init_hook=initializer, cache=CacheType.NO_CACHE) 时，训练集和测试集的 classification_error_evaluator 会与上述结论相差较大，分别为 9.xx% 和 20.xx%。

[官方文档](http://www.paddlepaddle.org/doc_cn/ui/data_provider/pydataprovider2.html#provider)中提到：
DataProvider提供了两种简单的Cache策略。他们是
CacheType.NO_CACHE 不缓存任何数据，每次都会从python端读取数据
CacheType.CACHE_PASS_IN_MEM 第一个pass会从python端读取数据，剩下的pass会直接从内存里 读取数据。

provider部分，还尝试了不同的 pool_size 和 min_pool_size 参数，都是这样类似的 classification_error_evaluator。

Cache 应该仅仅是缓存的策略，却对模型训练准确率产生影响。

**Comment by wangkuiyi** [comment]
@lazzylee 是不是这样描述更清楚一些：

**分类模型的准确率是不是不应当受 dataprovider 的缓存策略影响？**

我们做了一个简单的音频frame分类的试验。在实验中发现一些与预期不符合的情况。记录如下：

**试验配置**
- 语音的音频frame的分类；
- 训练集包括从1000句语音中提取的260,000个labeled vectors；测试集和训练集一样；
- 每个vector是484维；
- label set的大小是3513；
- 模型训练配置脚本源码在[这里](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/trainer_config.dnn.py)，使用了[Quick Start](http://www.paddlepaddle.org/doc_cn/demo/quick_start/index.html#optimization-algorithm) 文档中优化算法(Optimization Algorithm)部分的默认参数：
  - batch_size = 256, 
  - learning_rate=2e-3, 
  - learning_method=AdamOptimizer(), 
  - regularization=L2Regularization(8e-4), 
  - gradient_clipping_threshold=25
- dataprovider 源码在[这里](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/dataprovider_ark.py)

**试验过程**

我们尝试了修改 dataprovider 的 [Python provider decorator](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/dataprovider_ark.py#L27)：
1. 当 decorator 是`@provider(init_hook=initializer)`时，用 classification_error_evaluator 计算的训练集和测试集的正确率一致，都很低，大约 4%；
2. 当 decorator 是`@provider(init_hook=initializer, cache=CacheType.CACHE_PASS_IN_MEM)`时，准确率一致，且正常【是多少？】
3. 当 decorator 是`@provider(init_hook=initializer, cache=CacheType.NO_CACHE)` 时，训练集和测试集（一样）的准确率不一样，且相差较大，分别为 9.xx% 和 20.xx%。

**疑问**

Paddle的官方文档中提到：

> DataProvider提供了两种简单的Cache策略。他们是
> CacheType.NO_CACHE 不缓存任何数据，每次都会从python端读取数据
> CacheType.CACHE_PASS_IN_MEM 第一个pass会从python端读取数据，剩下的pass会直接从内存里 读取数据。

我们以为：在decorator中指定不同的缓冲策略，不应该影响到模型准确率？

此外，我们队试验3.中，同样的数据得到不同的结果，感到很纳闷。

**Comment by wangkuiyi** [comment]
其实我对于这个试验

> 当 decorator 是@provider(init_hook=initializer, cache=CacheType.NO_CACHE) 时，训练集和测试集（一样）的准确率不一样，且相差较大，分别为 9.xx% 和 20.xx%。

的结果很不解。

如果训练集和测试集是一样的，那么运行两次 classification_error_evaluator ，都用训练集，得到的准确率数字是一样的吗？

**Comment by lazzylee** [comment]
> 当 decorator 是@provider(init_hook=initializer, cache=CacheType.NO_CACHE) 时，训练集和测试集（一样）的准确率不一样，且相差较大，分别为 9.xx% 和 20.xx%。

这部分我也有些不解，我想可能有两方面原因
- 训练集和测试集虽然是一样的数据，但是可能读取方式，batch_size等存在差异。最终的 samples 数目存在轻微的差异，训练集要略少一些。
- 训练和测试时用的评价方式有稍许差别

我还没去细跟这个疑问
- 正常情况下，训练集和测试集的差异很微小。
- 训练集和测试集一致只是一个简单验证流程正确性的实验
- 训练集和测试集不一致时，准确率还很差，目前主要在跟踪这个问题

**Comment by lazzylee** [comment]
刚为了回答 @wangkuiyi  的疑问，我具体检查了训练集的测试集的 samples 数目。以下原因最后一句（最终的 samples 数目存在轻微的差异，训练集要略少一些），我没法重现了。

> 训练集和测试集虽然是一样的数据，但是可能读取方式，batch_size等存在差异。最终的 samples 数目存在轻微的差异，训练集要略少一些。

**Comment by wangkuiyi** [comment]
@lazzylee 能否把你的试验步骤都写成一个bash脚本之类的程序，这样可以复现你描述的问题，从而比较容易让paddle团队的人下手帮助定位和解决问题？

**Comment by lazzylee** [comment]
@wangkuiyi  大致步骤如下：
- [配置](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/trainer_config.dnn.py) 和 [dataprovider](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/dataprovider_ark.py)
- 首先，简单修改 data 和 demo 的 path (/home/core/paddle/data)，运行 [gpu-start-docker](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/gpu-start-docker.sh)
- 然后，运行 [gpu-train-in-docker](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/gpu-train-in-docker.sh)
- 最后，如上所述，修改  [dataprovider](https://github.com/unisound-ail/Paddle/blob/kaldi-dnn-demo/demo/kaldi_dnn/dataprovider_ark.py) 中的 @provider 

关于数据，现在有好的共享方式了吗？ @lipeng-unisound

**Comment by wangkuiyi** [comment]
了解了。上述comment里的内容合并一下，就能把这个问题说清楚了。

另外，这个issue里的代码是不是就在 https://github.com/unisound-ail/Paddle/pull/2 这里？

我想可以开一个PR，把 unisound-ail/Paddle 里的这个试验（实际上是一个新的demo）merge到baidu/Paddle里。这样baidu/Paddle 的成员就可以通过这个PR看到 @lazzylee 做的工作，并且提供帮助了。

**Comment by pineking** [comment]
@wangkuiyi 赞同，目前还是在 unisound-ail/Paddle repo 里

**Comment by lazzylee** [comment]
@wangkuiyi 赞同，但需要一个 数据下载/共享 的部分，才能算一个完整的 demo
